stdlib/internal/promql/join.go

```diff
@@ -0,0 +1,553 @@
+	if entry, ok := c.data.Delete(key); ok {
+		t := entry.(*cacheEntry)
+		if t.l != nil {
+			t.l.Done()
+		}
+		if t.r != nil {
+			t.r.Done()
+		}
+	}
+}
+
+func (c *mergeJoinCache) clean() {
+	var keys []flux.GroupKey
+	c.data.Range(func(key flux.GroupKey, value interface{}) {
+		keys = append(keys, key)
+	})
+	for _, key := range keys {
+		c.delete(key)
+	}
+}
+
+func (c *mergeJoinCache) Table(key flux.GroupKey) (flux.Table, error) {
+	entry, ok := c.data.Lookup(key)
+	if !ok {
+		return nil, errors.Newf(codes.NotFound, "no entry for group key %v in cache", key)
+	}
+	t := entry.(*cacheEntry)
+	if t.l == nil || t.r == nil {
+		return nil, errors.Newf(codes.NotFound, "no entry for group key %v in cache", key)
+	}
+	return c.joinTables(t.l, t.r)
+}
+
+func (c *mergeJoinCache) ForEach(f func(flux.GroupKey)) {
+	c.data.Range(func(key flux.GroupKey, value interface{}) {
+		if value.(*cacheEntry).ready() {
+			f(key)
+		}
+	})
+}
+
+func (c *mergeJoinCache) ForEachWithContext(f func(flux.GroupKey, execute.Trigger, execute.TableContext)) {
+	c.data.Range(func(key flux.GroupKey, value interface{}) {
+		if value.(*cacheEntry).ready() {
+			f(key, execute.NewTriggerFromSpec(c.spec), execute.TableContext{
+				Key: key,
+			})
+		}
+	})
+}
+
+func (c *mergeJoinCache) DiscardTable(key flux.GroupKey) {
+	c.delete(key)
+}
+
+func (c *mergeJoinCache) ExpireTable(key flux.GroupKey) {
+	c.delete(key)
+}
+
+func (c *mergeJoinCache) SetTriggerSpec(spec plan.TriggerSpec) {
+	c.spec = spec
+}
+
+func (c *mergeJoinCache) joinTables(a, b flux.Table) (flux.Table, error) {
+	// Compile row fn for the input rows
+	if err := c.fn.Prepare(a.Cols(), b.Cols()); err != nil {
+		return nil, err
+	}
+
+	key := a.Key()
+
+	// Create a table builder for the output of the join
+	builder := execute.NewColListTableBuilder(key, c.alloc)
+
+	ar := make([]flux.ColReader, 0)
+	br := make([]flux.ColReader, 0)
+
+	// Release column readers after joining
+	defer a.Done()
+	defer b.Done()
+
+	// Retain column readers for joining
+	if err := a.Do(func(cr flux.ColReader) error {
+		cr.Retain()
+		ar = append(ar, cr)
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	if err := b.Do(func(cr flux.ColReader) error {
+		cr.Retain()
+		br = append(br, cr)
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	ca := execute.ColIdx(execute.DefaultTimeColLabel, a.Cols())
+	if ca < 0 {
+		return nil, errors.New(codes.NotFound, "no _time column found")
+	}
+
+	iterA := &colReaderIterator{
+		i:       -1,
+		timeCol: ca,
+		readers: ar,
+	}
+
+	cb := execute.ColIdx(execute.DefaultTimeColLabel, b.Cols())
+	if cb < 0 {
+		return nil, errors.New(codes.NotFound, "no _time column found")
+	}
+
+	iterB := &colReaderIterator{
+		i:       -1,
+		timeCol: cb,
+		readers: br,
+	}
+
+	rowA := make(map[string]values.Value, 8)
+	rowB := make(map[string]values.Value, 8)
+
+	_, ra, ta := iterA.next()
+	i, rb, tb := iterB.next()
+
+	var firstRow = true
+
+NEXT:
+	if ra == -1 || rb == -1 {
```

Comment:
Can you encapsulate this section where it finds the same time into a function? Maybe into an `advance()` function and then document it to say "advance to the next row to be joined".

Reply:
Can you provide a bit more explanation?

Reply:
It seems to me like the purpose of this is to determine which index they have a matching time. It appears that this is implementing an inner join only, but I wasn't clear about that.

What I mean is that instead of using `NEXT` and a goto, this can be changed into a function called `advance()` that will call `next()` on each of the iterators until it finds a match. The inside implementation of that can use goto. I think goto is kind of awesome for this kind of thing. I just think it should be isolated to a single function so I know where it starts and where it ends.